plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -319,6 +319,138 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
+	GError *err_usb4_port_offline_path = NULL;
+	GError *err_usb4_port_rescan_path = NULL;
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, &err_usb4_port_offline_path)) {
+		if (!fu_udev_device_write_sysfs(device, offline, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting usb4 port offline failed");
+			return FALSE;
+		}
+	} else {
+		g_debug("%s", err_usb4_port_offline_path->message);
+		return TRUE;
+	}
+	if (fu_thunderbolt_check_usb4_port_path(device, rescan, &err_usb4_port_rescan_path)) {
+		if (!fu_udev_device_write_sysfs(device, rescan, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Rescan on port failed");
+			return FALSE;
+		}
+	} else {
+		g_debug("%s", err_usb4_port_rescan_path->message);
+		return TRUE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device, GError **err)
+{
+	FuUdevDevice *udev = FU_UDEV_DEVICE(device);
+	const gchar *offline = "usb4_port1/offline";
+	GError **err_usb4_port_offline_path = NULL;
+
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, err_usb4_port_offline_path)) {
+		if (!fu_udev_device_write_sysfs(udev, offline, "0", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting port online failed");
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static FuUdevDevice *
+fu_thunderbolt_get_udev_parent_device(FuDevice *device, GError **error)
+{
+	g_autoptr(GUdevDevice) udev_parent1 = NULL;
+	g_autoptr(GUdevDevice) udev_parent2 = NULL;
+	GUdevDevice *udev_device = NULL;
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "not a retimer device");
+		return NULL;
+	}
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	if (udev_device == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "failed to get udev device for retimer");
+		return NULL;
+	}
+	udev_parent1 = g_udev_device_get_parent(udev_device);
+	if (udev_parent1 == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "failed to get parent device for retimer");
+		return NULL;
+	}
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	if (udev_parent2 == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "failed to get host router device for retimer");
+		return NULL;
+	}
+	return fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+					       g_steal_pointer(&udev_parent2));
+}
+gboolean
+fu_thunderbolt_probe_retimer(FuDevice *device, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = NULL;
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
```

Comment:
can you flip these the other way around please? just by convention self is first.